pkg/resource/training_job/custom_delta.go

```diff
@@ -60,3 +65,53 @@ func customSetDefaults(
 		}
 	}
 }
+
+// SM returns profiler related objects even if the user disables the profiler
+// customPostCompare detects if there is a diff
+func customPostCompare(latest *resource, desired *resource, delta *ackcompare.Delta) {
+	profilerConfigDiff := delta.DifferentAt("Spec.ProfilerConfig")
+	profilerRuleDiff := delta.DifferentAt("Spec.ProfilerRuleConfigurations")
+	if !profilerConfigDiff && !profilerRuleDiff {
+		return
+	}
+	profilerStatus := latest.ko.Status.ProfilingStatus
+	profilerDisabled := false
+
+	if ackcompare.IsNotNil(profilerStatus) {
+		//Do not remove profiler if user wants to enable it
+		if *profilerStatus == "Disabled" && !userInitiatesProfilerCheck(desired) {
+			profilerDisabled = true
+		} else {
+			return
+		}
+	} else {
+		return
+	}
+	// TODO: Replace remove delta with an ack version when its natively supported
+	if profilerConfigDiff && profilerDisabled {
+		removeDelta(delta, "Spec.ProfilerConfig")
+	}
+	if profilerRuleDiff && profilerDisabled {
+		removeDelta(delta, "Spec.ProfilerRuleConfigurations")
+	}
+}
+
+// userInitiatesProfilerCheck checks if the user enabled/re enabled the profiler.
+func userInitiatesProfilerCheck(desired *resource) bool {
+	profilerConfigPresent := ackcompare.IsNotNil(desired.ko.Spec.ProfilerConfig)
+	profilerRuleConfigPresent := ackcompare.IsNotNil(desired.ko.Spec.ProfilerRuleConfigurations)
+	return profilerConfigPresent && profilerRuleConfigPresent
+}
+
+// removeDelta Removes fieldName from the delta slice.
+// TODO: Replace when ack runtime can do this.
+func removeDelta(delta *ackcompare.Delta, fieldName string) {
+	differences := delta.Differences
+	for index, diff := range differences {
+		if diff.Path.Contains(fieldName) {
+			differences = append(differences[:index], differences[index+1:]...)
```

Comment:
index + 1 can be out of bounds?

Reply:
If I have s [a] -> slice | cap = 1 length = 1
s[1] isnt valid because the length of the slice is 1. But s[1:] is valid because the capacity is one.
As per [go spec](https://go.dev/ref/spec#Slice_expressions) "For slices, the upper index bound is the slice capacity cap(a) rather than the length."
s[2:] would be invalid because 2 > cap(s)
Differences is a slice. cap(differences) = differences[index+1:]

Reply:
>is it better if we ignore the default generated delta comparison and use reflect.DeepEqual when needed to compare both?

delta.go already uses deep [equal](https://github.com/aws-controllers-k8s/sagemaker-controller/blob/1989765a2bccadb1200da8b8411e7eece000991f/pkg/resource/training_job/delta.go#L232), the objective of this code is to remove the elements in the delta if the profiler is disabled. The alternative would be manually overwriting the desired object in pre compare which would require more boilerplate and could potentially have edge cases.